File: python/scripts/predict.py
from __future__ import annotations

import argparse
from pathlib import Path

import numpy as np
import torch
from PIL import Image
from torchvision.transforms import functional as TF

from floodseg.model import build_model, forward_logits


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Predict a binary flood mask for one image.")
    parser.add_argument("image", help="Path to an input satellite/aerial image.")
    parser.add_argument("--checkpoint", default="outputs/checkpoints/best_model.pt")
    parser.add_argument("--image-size", type=int, default=256)
    parser.add_argument("--output", default="outputs/predictions/single_mask.png")
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    device = torch.device("cuda" if torch.cuda.is_available() else "mps" if torch.backends.mps.is_available() else "cpu")

    checkpoint = torch.load(args.checkpoint, map_location=device)
    model = build_model(pretrained=False).to(device)
    model.load_state_dict(checkpoint["model_state_dict"])
    model.eval()

    original = Image.open(args.image).convert("RGB")
    resized = TF.resize(original, [args.image_size, args.image_size], antialias=True)
    image_tensor = TF.to_tensor(resized).unsqueeze(0).to(device)

    with torch.no_grad():
        logits = forward_logits(model, image_tensor)
        prediction = (torch.sigmoid(logits)[0, 0].cpu().numpy() >= 0.5).astype(np.uint8) * 255

    output_mask = Image.fromarray(prediction).resize(original.size, resample=Image.Resampling.NEAREST)
    output_path = Path(args.output)
    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_mask.save(output_path)
    print(f"Saved predicted mask to: {output_path}")


if __name__ == "__main__":
    main()

File: R/scripts/predict.R
get_script_dir <- function() {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("^--file=", args, value = TRUE)
  if (length(file_arg) > 0) {
    return(dirname(normalizePath(sub("^--file=", "", file_arg[[1]]))))
  }
  getwd()
}

project_dir <- normalizePath(file.path(get_script_dir(), ".."))
local_lib <- file.path(project_dir, "packages")
if (dir.exists(local_lib)) {
  .libPaths(c(local_lib, .libPaths()))
}
torch_home <- file.path(project_dir, "torch-cache")
dir.create(torch_home, recursive = TRUE, showWarnings = FALSE)
Sys.setenv(TORCH_HOME = torch_home)
Sys.setenv(TORCH_CPP_LOG_LEVEL = "ERROR")
Sys.setenv(C10_LOG_LEVEL = "ERROR")
Sys.setenv(GLOG_minloglevel = "2")
Sys.setenv(PYTORCH_DISABLE_NNPACK = "1")
source(file.path(project_dir, "lib", "args.R"))
source(file.path(project_dir, "lib", "data.R"))
source(file.path(project_dir, "lib", "model.R"))
source(file.path(project_dir, "lib", "visualize.R"))

resolve_path <- function(path) {
  if (grepl("^/", path)) {
    return(path)
  }
  normalizePath(file.path(project_dir, path), mustWork = FALSE)
}

args <- parse_cli_args(list(
  "image" = "",
  "checkpoint" = "outputs/checkpoints/best_model.pt",
  "image-size" = "128",
  "base-channels" = "8",
  "output" = "outputs/predictions/single_mask.png"
))

if (args[["image"]] == "") {
  stop("Please provide an image path with --image path/to/image.jpg")
}

image_path <- resolve_path(args[["image"]])
checkpoint_path <- resolve_path(args[["checkpoint"]])
output_path <- resolve_path(args[["output"]])

device <- torch_device(if (cuda_is_available()) "cuda" else "cpu")
checkpoint <- torch_load(checkpoint_path)
checkpoint_args <- checkpoint$args
base_channels <- if (!is.null(checkpoint_args) && !is.null(checkpoint_args[["base-channels"]])) {
  as_int(checkpoint_args[["base-channels"]])
} else {
  as_int(args[["base-channels"]])
}
model <- unet_model(base_channels = base_channels)$to(device = device)
model$load_state_dict(checkpoint$model_state_dict)
model$eval()

image_tensor <- read_image_tensor(image_path, as_int(args[["image-size"]]))$unsqueeze(1)$to(device = device)
logits <- model(image_tensor)
predicted_mask <- (torch_sigmoid(logits[1, , , ]) >= 0.5)$to(dtype = torch_float())

save_mask_png(predicted_mask, output_path)
cat(sprintf("Saved predicted mask to: %s\n", output_path))
